core/src/banking_stage/tpu_to_pack.rs

```diff
@@ -0,0 +1,192 @@
+//! Service to send transaction packets to the external scheduler.
+//!
+
+use {
+    agave_banking_stage_ingress_types::BankingPacketReceiver,
+    agave_scheduler_bindings::{tpu_message_flags, SharableTransactionRegion, TpuToPackMessage},
+    rts_alloc::Allocator,
+    solana_packet::PacketFlags,
+    solana_perf::packet::PacketBatch,
+    std::{
+        net::IpAddr,
+        path::{Path, PathBuf},
+        ptr::NonNull,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+        thread::JoinHandle,
+    },
+};
+
+/// Spawns a thread to receive packets from TPU and send them to the external scheduler.
+///
+/// # Safety:
+/// - `allocator_worker_id` must be unique among all processes using the same allocator path.
+pub unsafe fn spawn(
+    exit: Arc<AtomicBool>,
+    non_vote_receiver: BankingPacketReceiver,
+    vote_receivers: Option<(BankingPacketReceiver, BankingPacketReceiver)>,
+    allocator_path: PathBuf,
+    allocator_worker_id: u32,
+    queue_path: PathBuf,
+) -> JoinHandle<()> {
+    std::thread::Builder::new()
+        .name("solTpu2Pack".to_string())
+        .spawn(move || {
+            // Setup allocator and queue
+            // SAFETY: The caller must ensure that no other process is using the same worker id.
+            if let Some((allocator, producer)) =
+                unsafe { setup(allocator_path, allocator_worker_id, queue_path) }
+            {
+                tpu_to_pack(exit, non_vote_receiver, vote_receivers, allocator, producer);
+            }
+        })
+        .unwrap()
+}
+
+fn tpu_to_pack(
+    exit: Arc<AtomicBool>,
+    non_vote_receiver: BankingPacketReceiver,
+    vote_receivers: Option<(BankingPacketReceiver, BankingPacketReceiver)>,
+    allocator: Allocator,
+    mut producer: shaq::Producer<TpuToPackMessage>,
+) {
+    // Create a round-robin receiver for vote/non-vote packets.
+    // If vote receivers are not provided, this just always returns the non-vote receiver.
+    let mut round_robin_receiver_iter = if let Some(vote_receivers) = vote_receivers {
+        vec![vote_receivers.0, vote_receivers.1, non_vote_receiver]
+            .into_iter()
+            .cycle()
+    } else {
+        vec![non_vote_receiver].into_iter().cycle()
+    };
+
+    while exit.load(Ordering::Relaxed) {
+        // Receive packets from the next receiver in round-robin order.
+        let Ok(packet_batches) = round_robin_receiver_iter.next().unwrap().try_recv() else {
+            continue;
+        };
+        handle_packet_batches(&allocator, &mut producer, packet_batches);
+    }
+}
+
+fn handle_packet_batches(
+    allocator: &Allocator,
+    producer: &mut shaq::Producer<TpuToPackMessage>,
```

Comment:
just think out loud - `allocator` and `producer` seems *always* have to work together. Maybe combine them into a single entity, say `SharedMemoryManager` (perhaps a bad name there :)), so it has: 
- `prep()` that does clean all remote frees, and sync prodcuer queue
- `allocate() -> (allocated_ptr, tpu_to_pack_message)` that does memory allocation for sending over packet

Reply:
I added a fn to do it that separates allocation/reservation into a fn that handles it correctly.

It also allowed easier separation of the message population fn which is now tested 😄 
that way we don't need to set up an allocator for every test, which is a bit annoying to do wrt cleaning up files etc